File: packages/ubaa_app/lib/src/calendar.dart
import 'package:ubaa_bindings/calendar.dart' as core;
import 'package:ubaa_domain/ubaa_domain.dart';
import 'package:ubaa_platform/ubaa_platform.dart';

BoyaCalendarActions createBoyaCalendarActions() {
  const platform = SystemCalendar();
  return BoyaCalendarActions(
    available: platform.available,
    draft: (course, reminder) {
      String? field(String name) => course.fields
          .where((field) => field.label == name)
          .firstOrNull
          ?.value;
      final draft = core.bykcCalendarDraft(
        title: course.title,
        location: field('地点'),
        start: field('开始'),
        end: field('结束'),
        selectStart: field('选课开始'),
        reminder: reminder,
      );
      if (draft == null) return null;
      return CalendarDraft(
        title: draft.title,
        location: draft.location,
        description: draft.description,
        startMs: draft.startMs.toInt(),
        endMs: draft.endMs.toInt(),
        reminderMinutes: draft.reminderMinutes,
      );
    },
    conflicts: (draft) async {
      final events = await platform.read(draft);
      return events
          .where(
            (event) => core.calendarOverlaps(
              start: draft.startMs,
              end: draft.endMs,
              otherStart: event.startMs,
              otherEnd: event.endMs,
            ),
          )
          .toList(growable: false);
    },
    edit: platform.edit,
  );
}

File: crates/ubaa-flutter-bridge/src/api/calendar.rs
//! Local calendar operations use only Core's public facade.
#[derive(Clone)]
pub struct BridgeCalendarDraft {
    pub title: String,
    pub location: String,
    pub description: String,
    pub start_ms: i64,
    pub end_ms: i64,
    pub reminder_minutes: Option<i32>,
}

#[flutter_rust_bridge::frb(sync)]
pub fn bykc_calendar_draft(
    title: String,
    location: Option<String>,
    start: Option<String>,
    end: Option<String>,
    select_start: Option<String>,
    reminder: bool,
) -> Option<BridgeCalendarDraft> {
    ubaa_core::facade::bykc_calendar_draft(
        &title,
        location.as_deref(),
        start.as_deref(),
        end.as_deref(),
        select_start.as_deref(),
        reminder,
    )
    .map(|draft| BridgeCalendarDraft {
        title: draft.title,
        location: draft.location,
        description: draft.description,
        start_ms: draft.start_ms,
        end_ms: draft.end_ms,
        reminder_minutes: draft.reminder_minutes,
    })
}

#[flutter_rust_bridge::frb(sync)]
pub fn calendar_overlaps(start: i64, end: i64, other_start: i64, other_end: i64) -> bool {
    ubaa_core::facade::calendar_overlaps(start, end, other_start, other_end)
}
